src/shared/schemas.ts

```diff
@@ -64,13 +68,31 @@ export class SchemaService {
             ])
     }
 
+    public isMapped(uri: vscode.Uri): boolean {
+        for (const h of this.handlers.values()) {
+            if (h.isMapped(uri)) {
+                return true
+            }
+        }
+        return false
+    }
+
     public async start(): Promise<void> {
         getDefaultSchemas(this.extensionContext).then(schemas => (this.schemas = schemas))
         await this.startTimer()
     }
 
-    public registerMapping(mapping: SchemaMapping): void {
+    /**
+     * Registers a schema mapping in the schema service.
+     *
+     * @param mapping
+     * @param flush Flush immediately instead of waiting for timer.
+     */
+    public registerMapping(mapping: SchemaMapping, flush?: boolean): void {
         this.updateQueue.push(mapping)
+        if (flush === true) {
```

Comment:
Feels leaky. We're adding a lot of logic to this module just so we can emit a telemetry metric somewhere else. Why not provide an easy way to emit the metric from a URI instead?

Reply:
This logic change removes unnecessary indeterminism in the way that the schema services works. Why does it have this async nature anyway?

Also this is the only compromise in the schema service. The rest of the logic changes in schemas.ts are to expose the service's understanding of "is this a filetype that we care about?"

> Why not provide an easy way to emit the metric from a URI instead?

Do you mean something like:

    schemas.isTracked(uri)

That's essentially what I've tried to get started here.

Reply:
> Why does it have this async nature anyway?

Definitely agree though that's another discussion. 

> Also this is the only compromise in the schema service. The rest of the logic changes in schemas.ts are to expose the service's understanding of "is this a filetype that we care about?"

I'm saying that we should let the schema service care about what it cares about instead of trying to extract it out. So like we packed a bunch of logic into this `onDidOpenTextDocument` listener. What if we extracted it out and let relevant callers provide the information rather than the other way around?

In theory, putting this metric under a single event is a good idea. And it would work well if we didn't need a bunch of context. But in practice we clearly need a lot of information about why this document was opened, forcing us to expose the information from different systems. So instead we should generalize the 'throttling' behavior but still let callers provide the context.